jmdaemon/jmdaemon/onionmc.py

```diff
@@ -1084,10 +1111,27 @@ def process_control_message(self, peerid: str, msgtype: int,
                 return True
             try:
                 peerlist = msgval.split(",")
-                for peer in peerlist:
+                for peer_in_list in peerlist:
+                    # directories should send us peerstrings that include
+                    # nick;host:port;D where "D" indicates that the directory
+                    # is signalling this peer as having left. Otherwise, without
+                    # the third field, we treat it as a "join" event.
+                    try:
+                        nick, hostport, disconnect_code = peer_in_list.split(
+                            NICK_PEERLOCATOR_SEPARATOR)
+                        if not disconnect_code == "D":

```

Comment:
Will edit.

Re: just `continue`-ing instead of raising an error if not "D", I was doing that as a forwards compatibility thing, so that if we choose to add a different code there, old versions will just ignore the message.